### PR TITLE
Select v7

### PR DESCRIPTION
**Pull request by nickwmurphy**
### Summary
I didn't set out to make breaking changes, but in an effort to update `titleValue` when the `titleValue` prop changed, this seemed to be a fix for many of the remaining oddities of this particle.

1. `clearable` prop added: previously, `updateOnSelect` determined whether an instance of Select was able to be cleared. This new boolean prop replaces that.

2. **MOST NOTABLE CHANGE**: `titleValue` no longer controlled by internal state: the "title value" will now be controlled by props, not a mix of props and internal state
 
3. `updateOnSelect` prop removed: now that `titleValue` is controlled outside the component, there is no need for this. 

4. `valueFunction` prop removed: now that `titleValue` is controlled outside the component, there is no need for this.

The sandbox is now a better example (IMO) of how Select is most effectively implemented in a client.

### CODE REVIEW

**Comment by nickwmurphy** [comment]
Removed that test, as it was no longer necessary with these changes. I'll work with @ben-burlingham to see what test coverage is missing.

File: select/test/Select.test.jsx [hunk 1]
```diff
@@ -68,16 +68,6 @@ describe('Select', () => {
         expect(textElement.text()).toEqual(titleValue);
       });
 
-      it('can be cleared', () => {
```
Comment (mcclayton) [inline select/test/Select.test.jsx]: Isn't this test still relevant? Can we fix it instead of removing it?

File: select/test/Select.test.jsx [hunk 2]
```diff
@@ -106,17 +96,28 @@ describe('Select', () => {
...
-        const wrapper = shallowRender({ titleValue });
-        const clearButtonElement = wrapper.find(`.${baseStyles.clear}`);
-        expect(clearButtonElement.exists()).toBe(true);
-      });
-
-      it('does not render a clear button if there is no title value', () => {
-        const wrapper = shallowRender();
-        const clearButtonElement = wrapper.find(`.${baseStyles.clear}`);
-        expect(clearButtonElement.exists()).toBe(false);
+      describe('has title value', () => {
```
Comment (mcclayton) [inline select/test/Select.test.jsx]: `nb`
Maybe rephrase these two outer describes to be 
`when title value present`
and
`when title value not present`

File: select/src/Select.jsx [hunk 3]
```diff
@@ -278,7 +268,8 @@ export default class Select extends React.Component {
 
     const title = (
       <div className={cx(this.styles.staticText,
```
Comment (mcclayton) [inline select/src/Select.jsx]: `nb`
Minor: but this can be:
```
<div className={cx(this.styles.staticText, { 
        [this.styles.noClear]: !clearable,
        [this.styles.placeholder]: !titleValue 
})}>
```
You don't need to wrap each conditional style in an individual object. It can simply be one single object passed to classnames package.

File: select/src/Dropdown.jsx [hunk 4]
```diff
@@ -136,13 +138,21 @@ export default class Dropdown extends React.Component {
 
   onClearAll = () => {
     this.props.onClearAll();
```
Comment (mcclayton) [inline select/src/Dropdown.jsx]: `onClearAll` is not a required prop and the default is `null` yet there is no safeguard here? This should either handle the null case or it should be a required prop.

Comment (mcclayton) [reply select/src/Dropdown.jsx]: Looks like this is the case for basically all of the function type props, these should be handled as well.

File: select/src/Select.jsx (hunk 3 above)
Comment (nickwmurphy) [reply select/src/Select.jsx]: Fixed 👍

File: select/src/Dropdown.jsx (hunk 4 above)
Comment (ben-burlingham) [reply select/src/Dropdown.jsx]: @nickwmurphy lmk if you want me to help out here

Comment (ben-burlingham) [reply select/src/Dropdown.jsx]: That seems correct - should put in `() => {}` or similar.  But it occurs to me that this has been this way for a while and not had an issue, so React must handle it cleanly?  Anyways, `() => {}` is more correct :)

Comment (ben-burlingham) [reply select/src/Dropdown.jsx]: Ahh took a look at this with Nick. Seems there's several places where a prop like that is called, but there's a check to see if it's null or not.  The above case is indeed a danger point.  We'll just put in `() => {}` as defaults and remove the checks everywhere.

Nice catch!

Comment (mcclayton) [reply select/src/Dropdown.jsx]: 👍
